File: BalancedParentheses/BalancedParentheses.cpp
// BalancedParentheses.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <string>
#include <stack>

bool IsBalancedParentheses(std::string& str)
{
    std::stack<char> s;
    for (auto c : str)
    {
        switch (c)
        {
        case '(':
            s.push(')');
            break;

        case '{':
            s.push('}');
            break;

        case '[':
            s.push(']');
            break;

        case ')':
        case '}':
        case ']':
            if (s.empty() || s.top() != c)
                return false;
            s.pop();
            break;

        default:
            break;
        }
    }

    return true;
}

int main()
{
    while (true)
    {
        std::cout << "Please enter a string (\"q\" to exit): ";
        std::string str;
        std::cin >> str;
        if (str == "q" || str == "Q")
            break;

        std::cout << "String \"" << str << "\" " << (IsBalancedParentheses(str) ? "is" : "is not") << " Parentheses balanced" << std::endl;
    }
}

File: Go/BalancedParentheses.go
package main

import "fmt"

type Stack struct {
	size  int
	stack []byte
}

func NewStack() *Stack {
	return &Stack{0, make([]byte, 10000)}
}

func (s *Stack) IsEmpty() bool {
	return s.size == 0
}

func (s *Stack) Size() int {
	return s.size
}

func (s *Stack) Push(b byte) {
	s.stack[s.size] = b
	s.size++
}

func (s *Stack) Pop() {
	if s.size == 0 {
		return
	}

	s.size--
}

func (s *Stack) Top() byte {
	if s.size == 0 {
		return 0
	}

	return s.stack[s.size-1]
}

func IsBalancedParentheses(str string) bool {
	s := NewStack()
	for _, c := range str {
		switch c {
		case '(':
			s.Push(')')

		case '{':
			s.Push('}')

		case '[':
			s.Push(']')

		default:
			if c != rune(s.Top()) {
				return false
			}
			s.Pop()
		}
	}

	return s.IsEmpty()
}

func main() {
	for {
		fmt.Print("Please enter the string: ")
		var str string
		fmt.Scan(&str)
		if str == "q" || str == "Q" {
			break
		}

		b := IsBalancedParentheses(str)
		fmt.Print("The string is ")
		if !b {
			fmt.Print("not ")
		}
		fmt.Println("balenced!")
	}
}
